### PR TITLE
Urn display

### PR DESCRIPTION
**Pull request by nicpottier**

### CODE REVIEW

File: temba/channels/types/twitter/tasks.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+from temba.utils import chunk_list
+from django.conf import settings
+
+
+@task(track_started=True, name='resolve_twitter_ids_task')
+def resolve_twitter_ids():
+    r = get_redis_connection()
+
+    # TODO: we can't use our non-overlapping task decorator as it creates a loop in the celery resolver when registering
+    if r.get('resolve_twitter_ids_task'):  # pragma: no cover
```
Comment (rowanseymour) [inline temba/channels/types/twitter/tasks.py]: Did you figure out why that happens for this task but not others?

Comment (nicpottier) [inline temba/channels/types/twitter/tasks.py]: Others don't get registered dynamically like this, so this is the only example.

Comment (rowanseymour) [inline temba/channels/types/twitter/tasks.py]: Maybe this is the problem https://stackoverflow.com/a/41335949? (`nonoverlapping_task` creates a `shared_task`)

Comment (nicpottier) [inline temba/channels/types/twitter/tasks.py]: I fought that problem separately.. the behavior with this one is actually a total hang of Celery, I pdb'ed it a bit and strangely it does it when trying to check type of the task with an instanceof. Don't know if there is some kind of circular aspect to it perhaps that then isn't caught? Celery likes failing silently.

File: temba/contacts/models.py [hunk 2]
```diff
@@ -95,12 +95,18 @@ def from_parts(cls, scheme, path):
         if not path:
             raise ValueError("Invalid path component: '%s'" % path)
 
-        return '%s:%s' % (scheme, path)
+        urn = '%s:%s' % (scheme, path)
+
+        if display is not None:
```
Comment (rowanseymour) [inline temba/contacts/models.py]: I don't think we want to allow empty strings as display

Comment (nicpottier) [inline temba/contacts/models.py]: Ya, it is a bit weird. The idea is that we could then "know" if we'd tried to look up the display for a URN. An empty display means we didn't find it or there isn't one, while None means we haven't tried to look it up.

A bit subtle but prevents us from looking things up repeatedly.. what do you think?

Comment (rowanseymour) [inline temba/contacts/models.py]: Can we use the db value of ContactURN.display instead? I guess it's not much better but I think I'm more ok with using that in a subtle way than passing around URNs ending with `#`

Comment (nicpottier) [inline temba/contacts/models.py]: Don't we want to keep those mapping 1-1? Seems like a big source of possible errors if we lose information in the string/unstring marshalling.

Comment (rowanseymour) [inline temba/contacts/models.py]: Fair point. What about a new boolean field on URN? That would be the thing we index on then to determine which of an org's URNs need looked-up. I know it's yet-another-field but means we get to keep URNs which look like they follow an RFC

Comment (nicpottier) [inline temba/contacts/models.py]: But then you lose the ability to communicate in string form whether the URN has had that display looked up or not. Do you just not like the look of it?

Comment (rowanseymour) [inline temba/contacts/models.py]: I'm not convinced that's information that should be encoded in the string version - it's not useful information outside of RapidPro. Certainly not every field on the `ContactURN` model belongs in the string representation, e.g. preferred channel, priority

Comment (nicpottier) [inline temba/contacts/models.py]: Ok. I would rather just continue retrying ones we can't resolve then add another field to URN though, so I'll just go that route.

Comment (rowanseymour) [inline temba/contacts/models.py]: Is it really so bad to add a boolean field? We're gonna need an index of URNs that need processed regardless

Comment (nicpottier) [inline temba/contacts/models.py]: We need an index on org/scheme/display where display is not null, but I think that's it.

In the end not trying to refetch these is an optimization and I'd rather not add extra random fields for particular cases unless we see that as an actual performance problem.

Comment (nicpottier) [inline temba/contacts/models.py]: Blech.. this stuff is a mess. So upon looking at this a bit more closely I think we DO want a string representation that helps us determine whether the path of a URN is a user_id or a screen_name.

Essentially URNs like `twitter:nicpottier` are going to continue to exist in the system and be imported or set through our API. And those URNs may actually be `twitter:12345` which would be pointing to the screen name 12345, not the twitter user with user_id 12345.

So I think we do need to be able to say that `twitter:12345` is different than `twitter:12345#`. The latter means the twitter user with user_id `12345` the former means the twitter user with screen name `12345`.

If we do that, then we can keep things consistent in how we deal with URNs through APIs and the like, things will continue to work as they do today and we can do properly look things up. If we don't, then suddenly Twitter users that have numerical screen names are totally borked.

The only other way I can think of dealing with this is to introduce a different scheme entirely, but that seems worse than having a `#` indicate the difference between displays and internal ids. Do agree it is subtle and weird though.. maybe a call tomorrow morning to deathmatch over it?

Comment (rowanseymour) [inline temba/contacts/models.py]: See you tomorrow 💀

**Comment by nicpottier** [comment]
Closing this as we are going `twitterid` scheme route.

**Pull request closed without merging**